File: frontend/src/types/product.ts
import type {ReviewResponse} from './review';
export interface CategoryResponse {
  id?: number;
  name: string;
}

export interface ProductVariantRequest {
  productId: number;
  sku: string;
  price: string;
  originalPrice?: number;
  stock: number;
  attributes?: Record<string, string>;
}

export interface ProductVariantResponse extends ProductVariantRequest {
  id: number; 
}

export interface ProductRequest {
  name: string;
  images: string[];
  poster: string;
  description?: string;
  categoryId: number;
  specifications?: Record<string, string>;
  variants?: ProductVariantRequest[]; 
}

export interface ProductResponse extends Omit<ProductRequest, 'categoryId' | 'variants'> {
  price: number;
  id: number;
  category: CategoryResponse;
  variants?: ProductVariantResponse[];
  reviews?: ReviewResponse[];
  averageRating?: number;
  totalReviews?: number;
  createdAt?: string;
}

File: backend/src/main/java/com/tientoan21/entity/Product.java
package com.tientoan21.entity;

import com.tientoan21.enums.ProductStatus;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.Formula;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Builder
@EqualsAndHashCode(callSuper = true)
@Entity
@Data
@AllArgsConstructor
@NoArgsConstructor
@Table(name = "products")
public class Product extends BaseEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String name;
    private String description;

    @ElementCollection
    private List<String> images;
    private String poster;
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    private Category category;

    @ElementCollection
    @CollectionTable(name = "product_specs")
    @MapKeyColumn(name = "spec_key")
    @Column(name = "spec_value")
    private Map<String, String> specifications = new HashMap<>();

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL,orphanRemoval = true)
    private List<ProductVariant> variants = new ArrayList<>();

    @Formula("(SELECT COALESCE(AVG(r.rating), 0.0) FROM reviews r WHERE r.product_id = id AND r.deleted_at IS NULL)")
    private Double averageRating;

    @Formula("(SELECT COUNT(*) FROM reviews r WHERE r.product_id = id AND r.deleted_at IS NULL)")
    private Integer totalReviews;
}
